Improve search behaviour for partial words

dylanmccall opened the pull request:
This change breaks up raw and spell-corrected queries into two clauses,
each with FEATURE_PARTIAL, which are ORed together. It is necessary to
build them like this because FEATURE_PARTIAL does not work with a clause
which itself consists of boolean operators.

https://phabricator.endlessm.com/T27383

dylanmccall committed:
Improve search behaviour for partial words

This change breaks up raw and spell-corrected queries into two clauses,
each with FEATURE_PARTIAL, which are ORed together. It is necessary to
build them like this because FEATURE_PARTIAL does not work with a clause
which itself consists of boolean operators.

https://phabricator.endlessm.com/T27383

dylanmccall commented:
Here are some examples to test this in a shell for com.endlessm.astronomy.en:

    kermit query com.endlessm.astronomy.en "medie"
    kermit query com.endlessm.astronomy.en "mediev"
    kermit query com.endlessm.astronomy.en "medieval"

If you run that command in a runtime containing the proposed version of libdmodel, you should get the same result (_"Astronomy in the medieval Islamic world"_) for all three queries. In the current stable runtime, com.endlessm.apps.Sdk//5, the first query will return nothing.

dylanmccall commented:
The changes I made to testQuery.js are probably illustrative of the differently-behaved searches with this branch.

### beatles -> beetles

#### Old

    Query(
        (
            (
                (
                    XEXACTSbeatles@1 OR (Sbeatles@1 OR Sbeetles@2)
                ) OR (
                    (SYNONYM WILDCARD OR beatles) OR beatles@1
                )
            ) OR (
                (SYNONYM WILDCARD OR beetles) OR beetles@1
            )
        )
    )

#### New

    Query(
        (
            (
                (
                    XEXACTSbeatles@1 OR (
                        (
                            (SYNONYM WILDCARD OR Sbeatles) OR Sbeatles@1
                        ) OR (
                            (SYNONYM WILDCARD OR Sbeetles) OR Sbeetles@1
                        )
                    )
                ) OR (
                    (SYNONYM WILDCARD OR beatles) OR beatles@1
                )
            ) OR (
                (SYNONYM WILDCARD OR beetles) OR beetles@1
            )
        )
    )

### happy nwe year -> happy new year

#### Old

    Query(
        (
            XEXACTShappy_nwe_year@1 OR (
                (
                    Shappy@1 OR (
                        Snwe@2 OR Snew@3
                    )
                ) OR (
                    (
                        SYNONYM WILDCARD OR Syear
                    ) OR Syear@4
                )
            )
        )
    )

#### New

    Query(
        (
            XEXACTShappy_nwe_year@1 OR (
                (
                    (
                        Shappy@1 OR Snwe@2
                    ) OR (
                        (
                            SYNONYM WILDCARD OR Syear
                        ) OR Syear@3
                    )
                ) OR (
                    (
                        Shappy@1 OR Snew@2
                    ) OR (
                        (
                            SYNONYM WILDCARD OR Syear
                        ) OR Syear@3
                    )
                )
            )
        )
    )

The gist is we're now querying "[original search] OR [corrected search]", rather than interspersing a correction for each term. Alas, makes for some longer queries, but I _think_ it makes sense.

ptomato commented on dmodel/dm-query.c:
```diff
@@ -577,51 +576,66 @@ get_title_clause (XapianQueryParser *qp,
                                : 0;
 
-  guint max_length = MAX (terms_length, corrected_terms_length);
+  g_auto(GStrv) parser_terms = g_new0 (gchar *, terms_length + 1);
```
The code as it's written now looks like it might not need to duplicate `terms` and `corrected_terms`, since they only need to be passed to `g_strjoinv()`.

dylanmccall replied:
Oh, right! I'll put back that deduplication, actually, and indeed only need to use g_new0 the one time (for the deduplicated corrected_terms strv).

ptomato reviewed (changes requested):
There's a small improvement possible as far as I can see.

ptomato commented:
So I can see the improvement in the "beatles" case, which was that we didn't do the partial expansion in the synopsis search. But in the "happy nwe year" case we are adding no new terms to the search, but duplicating the already existing ones. Is that because the SYNONYM WILDCARD terms were actually not working as advertised?

### beatles -> beetles
Here we add the following new search terms compared to the original query:
```
(SYNONYM WILDCARD OR Sbeatles)
(SYNONYM WILDCARD OR Sbeetles)
```
### happy nwe year -> happy new year
No new search terms, although these are added to the query as duplicates:
```
Syear
Shappy
(SYNONYM WILDCARD OR Syear)
```
(in the old code we did remove corrected terms that were equal to the original term, but I'm not sure whether that's possible in the new code)

Maybe it would be helpful to see the difference between `medie`, `mediev`, and `medieval` written out in the kind of query diagrams that you posted above :smile:

dylanmccall commented:
Okay, put back the deduplication behaviour, and I noticed an issue where I was adding XAPIAN_QUERY_PARSER_FEATURE_PARTIAL for the _corrected_ terms, which wouldn't make much sense. The resulting queries with that adjustment seem more logical :) Good idea to generate those other examples! I'll do that shortly. For the ones from the test suite, they're somewhere in between the old and new I posted above…

dnohales commented on dmodel/dm-query.c:
```diff
@@ -577,51 +576,60 @@ get_title_clause (XapianQueryParser *qp,
+  for (guint i = 0; i < corrected_terms_length; i++)
+    {
+      gchar *corrected_term = corrected_terms[i];
+      if (g_strv_contains(terms, corrected_term) == FALSE) {
```
The indentation style is wrong.

dnohales reviewed (commented):
I'm not too familiar with Xapian querying to be honest (for example, I don't get why the synonym wildcard is only used for the "year" word). So I've just put an irrelevant comment about code style :P

dylanmccall committed:
Remove unnecessary strv duplication in get_title_clause

In addition, only use partial matching for the original input, not for
the corrected terms.

https://phabricator.endlessm.com/T27383

dylanmccall commented:
Okay, new new query examples. I actually ended up taking that deduplication code out again, so there's still a bit of duplication happening here but the output and the code are both tidier than my original solution, and I assume we can trust Xapian to do a better job handling duplication than we will. I'm a little afraid of how that might affect affect weighting of results, but initial poking and prodding (and automated tests) suggest that it's fine.

### beatles -> beetles

#### Old

    Query(
        (
            (
                (
                    XEXACTSbeatles@1 OR (Sbeatles@1 OR Sbeetles@2)
                ) OR (
                    (SYNONYM WILDCARD OR beatles) OR beatles@1
                )
            ) OR (
                (SYNONYM WILDCARD OR beetles) OR beetles@1
            )
        )
    )

#### New

    Query(
        (
            (
                (
                    XEXACTSbeatles@1 OR (
                        (
                            (SYNONYM WILDCARD OR Sbeatles) OR Sbeatles@1
                        ) OR Sbeetles@1
                    )
                ) OR (
                    (SYNONYM WILDCARD OR beatles) OR beatles@1
                )
            ) OR (
                (SYNONYM WILDCARD OR beetles) OR beetles@1
            )
        )
    )

### happy nwe year -> happy new year

#### Old

    Query(
        (
            XEXACTShappy_nwe_year@1 OR (
                (
                    Shappy@1 OR (
                        Snwe@2 OR Snew@3
                    )
                ) OR (
                    (
                        SYNONYM WILDCARD OR Syear
                    ) OR Syear@4
                )
            )
        )
    )

#### New

    Query(
        (
            XEXACTShappy_nwe_year@1 OR (
                (
                    (Shappy@1 OR Snwe@2) OR (
                        (SYNONYM WILDCARD OR Syear) OR Syear@3
                    )
                ) OR (
                    Shappy@1 OR Snew@2 OR Syear@3
                )
            )
        )
    )

### medie (com.endlessm.astronomy.en)

#### Old

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmedie) OR ZXEXACTSmedi@1
            ) OR (ZSmedi@1 OR ZSmedia@2)
        )
    )

#### New

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmedie) OR ZXEXACTSmedi@1
            ) OR (
                (
                    (SYNONYM WILDCARD OR Smedie) OR ZSmedi@1
                ) OR ZSmedia@1
            )
        )
    )

### mediev (com.endlessm.astronomy.en)

#### Old

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmediev) OR ZXEXACTSmediev@1
            ) OR (ZSmediev@1 OR ZSmedea@2)
        )
    )

#### New

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmediev) OR ZXEXACTSmediev@1
            ) OR (
                (
                    (SYNONYM WILDCARD OR Smediev) OR ZSmediev@1
                ) OR ZSmedea@1
            )
        )
    )

### medieval (com.endlessm.astronomy.en)

#### Old

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmedieval) OR ZXEXACTSmediev@1
            ) OR (ZSmediev@1 OR ZSmedeiv@2)
        )
    )

#### New

    Query(
        (
            (
                (SYNONYM WILDCARD OR XEXACTSmedieval) OR ZXEXACTSmediev@1
            ) OR (
                (
                    (SYNONYM WILDCARD OR Smedieval) OR ZSmediev@1
                ) OR ZSmedeiv@1
            )
        )
    )

dnohales reviewed (approved):
Looks way more reasonable and simpler compared to the old code, makes me think of a what's the reason the old code was that kind of complicated. Anyway.

If your manual tests works, it's fine to me.

dylanmccall commented:
Cool beans! I'll merge this to master so we can do some wider testing, but feel free to leave notes here if you spot any issues :)
